### ISSUE TITLE
Fetch Orders [API]

### ISSUE THREAD

**Issue by Morph21** [opencart/opencart#11760]
**What version of OpenCart are you reporting this for?**

**4.0.1.1**

**Describe the bug**
I'm trying to fetch orders/order with api

I did try url
/index.php?route=api/sale/order&api_token=[api/account/login response token]&order_id=2
/index.php?route=api/sale/order&api_token=[api/account/login response token]
/index.php?route=api/sale/order/info&api_token=[api/account/login response token]&order_id=2
/index.php?route=api/sale/order/history&api_token=[api/account/login response token]

Result is always:  ```404 The page you requested cannot be found!```

**Expected behavior**
I wish that would work.

**Comment by Morph21** [comment]
Second thing is that there is no method in catalog/controller/api/sale/order.php which would return an order for specific "id".
method "load" returns only "success" message in json

Next thing is that there is no method there which can use getOrders() method from /catalog/model/account/order.php

**Comment by mhcwebdesign** [comment]
Getting the order_info via an API call isn't needed in OpenCart 4, it's controllers normally use something like `$this->model_checkout_order->getOrder( $order_id  )`
 
What exactly are you trying to accomplish?

**Comment by osworx** [comment]
> Getting the order_info via an API call isn't needed in OpenCart 4, it's controllers normally use something like `$this->model_checkout_order->getOrder( $order_id )`
> 
> What exactly are you trying to accomplish?

Why should it be needed?
E.g. having an external script and I want to get infos about an order.
If the API would provide such a function, the call can be easy achieved.

And this is what (I think) @Morph21 want to do .. or not?

In general OC should provide a robust API (interface) which can be used by external scripts and programs.

**Comment by Morph21** [comment]
In general WEB api should provide you an option to get data only by using API.

So if I want to fetch orders for my external service (Accounting purposes) I can't do that now. 
It should be possible to download orders using API then you should be able to download details of each order by id.
Ideally it should be possible to fetch an list of orders with one call like it's done in shoplo, prestashop, woocommerce and many many other ecommerce softwares.

In my case I'm doing API integrations with a lot of ecommerce websites for ours website, we have several thousands of clients and many of them have ecommerce websites hosted on their own so I can't just modify php file for everyone

**Comment by Morph21** [comment]
> > Getting the order_info via an API call isn't needed in OpenCart 4, it's controllers normally use something like `$this->model_checkout_order->getOrder( $order_id )`
> > What exactly are you trying to accomplish?
> 
> Why should it be needed? E.g. having an external script and I want to get infos about an order. If the API would provide such a function, the call can be easy achieved.
> 
> And this is what (I think) @Morph21 want to do .. or not?
> 
> In general OC should provide a robust API (interface) which can be used by external scripts and programs.

@osworx exactly that, I need informations like total price of order, price per item row, quantity, address for invoice and delivery, taxes, delivery price and so on just to create an invoice. And while I have complex system for managing a lot of users I just need to get orders from OC.

Right now getting order by id isn't working in OC 4.0+, it works in 3.0.3.8. So assuming that it will start to work in 4.0 my best way to get all orders would be to iterate from 1 to infinity and each time fetching one order (or zero if someone deletes one)

**Comment by Morph21** [comment]
And about error itself 

You can see in screens below that in version 4.0 you have only one line adding something to json output which is this
`$json['success'] = $this->language->get('text_success');`

On the other hand in version 3.0+ you have this 2 lines of code

```
$json['order'] = $order_info;
$json['success'] = $this->language->get('text_success');
```

Here is a code from version 4.0+

![image](https://user-images.githubusercontent.com/38075691/192824825-bd94ab16-f0a5-4bbf-b9d2-aaca84dc4654.png)


Here is a code from version 3.0+
![image](https://user-images.githubusercontent.com/38075691/192825186-a5f9a97c-dd3f-4200-871a-c7e0b393bcf8.png)


In summary

**1. There is an error while fetching order by ID in version 4.0 - you can see that function is fetching order but not adding it to resposne**
**2. There should be a way to fetch list of orders filtered by some date/offset/limit or whatever to be able to do anything with orders**

Even if I would like to create an single order with API then I would like to fetch it to show it to my user.

**Comment by mhcwebdesign** [comment]
Be it as it is, this is not an OpenCart bug. The catalog/controller/api/* classes are for OpenCart only, shared by the catalog frontend and the admin backend. It's not meant to be a RESTful API.

However, it shouldn't be too hard to implement a proper RESTful API for OpenCart as an extension.

**Comment by osworx** [comment]
As far as I know, there is one API: https://github.com/opencartbrasil/opencart-rest-api
Maybe also more ..

**Comment by Morph21** [comment]
> Be it as it is, this is not an OpenCart bug. The catalog/controller/api/* classes are for OpenCart only, shared by the catalog frontend and the admin backend. It's not meant to be a RESTful API.
> 
> However, it shouldn't be too hard to implement a proper RESTful API for OpenCart as an extension.

Now I get it, btw you should kinda mention about that somewhere In your documentation
https://docs.opencart.com/en-gb/system/users/api/
which btw is outdated

**Comment by danielkerr** [comment]
https://github.com/opencart/opencart/issues/11438
